ethers-providers/src/transports/ws.rs

```diff
@@ -189,101 +205,135 @@ where
-            complete => {},
-        };
+    // dispatch an RPC request
+    async fn service_request(
+        &mut self,
+        id: u64,
+        request: String,
+        sender: Pending,
+    ) -> Result<(), ClientError> {
+        if self.pending.insert(id, sender).is_some() {
+            warn!("Replacing a pending request with id {:?}", id);
+        }
+
+        if let Err(e) = self.ws.send(Message::Text(request)).await {
+            error!("WS connection error: {:?}", e);
+            self.pending.remove(&id);
+        }
+        Ok(())
+    }
 
+    /// Dispatch a subscription request
+    async fn service_subscribe(&mut self, id: U256, sink: Subscription) -> Result<(), ClientError> {
+        if self.subscriptions.insert(id, sink).is_some() {
+            warn!("Replacing already-registered subscription with id {:?}", id);
+        }
         Ok(())
     }
 
-    async fn handle_request(&mut self, msg: TransportMessage) -> Result<(), ClientError> {
-        match msg {
-            TransportMessage::Request {
+    /// Dispatch a unsubscribe request
+    async fn service_unsubscribe(&mut self, id: U256) -> Result<(), ClientError> {
+        if self.subscriptions.remove(&id).is_none() {
+            warn!(
+                "Unsubscribing from non-existent subscription with id {:?}",
+                id
+            );
+        }
+        Ok(())
+    }
+
+    /// Dispatch an outgoing message
+    async fn service(&mut self, instruction: Instruction) -> Result<(), ClientError> {
+        match instruction {
+            Instruction::Request {
                 id,
                 request,
                 sender,
-            } => {
-                if self.pending.insert(id, sender).is_some() {
-                    warn!("Replacing a pending request with id {:?}", id);
-                }
+            } => self.service_request(id, request, sender).await,
+            Instruction::Subscribe { id, sink } => self.service_subscribe(id, sink).await,
+            Instruction::Unsubscribe { id } => self.service_unsubscribe(id).await,
+        }
+    }
 
-                if let Err(e) = self.ws.send(Message::Text(request)).await {
-                    error!("WS connection error: {:?}", e);
-                    self.pending.remove(&id);
-                }
-            }
-            TransportMessage::Subscribe { id, sink } => {
-                if self.subscriptions.insert(id, sink).is_some() {
-                    warn!("Replacing already-registered subscription with id {:?}", id);
+    async fn handle_ping(&mut self, inner: Vec<u8>) -> Result<(), ClientError> {
+        self.ws.send(Message::Pong(inner)).await?;
+        Ok(())
+    }
+
+    async fn handle_text(&mut self, inner: String) -> Result<(), ClientError> {
+        let message = serde_json::from_str::<Incoming>(&inner);
+        if message.is_err() {
+            // Ignore malformatted server responses
+            return Ok(());
+        }
+
+        match message.unwrap() {
```

Comment:
You can do `match message { Ok(Incoming::Response(resp) => { ...}, Ok(Incoming::Notification(notification) }` etc. if you want to avoid the double nesting of `match` and not use `.unwrap()`

Reply:
dbf6de0